stl-decomp-4j/src/main/java/com/github/servicenow/ds/stats/stl/SeasonalTrendLoess.java

```diff
@@ -715,18 +741,26 @@ public void smoothSeasonal(int width, boolean restoreEndPoints) {
 	 */
 	public Decomposition decompose() {
 		// TODO: Pass input data to decompose and reallocate buffers based on that size.
-
-		fDecomposition = new Decomposition(fData);
+		fDecomposition = new Decomposition(fData, fExogenousData);
 
 		int outerIteration = 0;
 		while (true) {
 
 			boolean useResidualWeights = outerIteration > 0;
 
-			for (int iteration = 0; iteration < fInnerIterations; ++iteration) {
-				smoothSeasonalSubCycles(useResidualWeights);
-				removeSeasonality();
-				updateSeasonalAndTrend(useResidualWeights);
+			if (fExogenousData == null) {
+				for (int iteration = 0; iteration < fInnerIterations; ++iteration) {
+					smoothSeasonalSubCycles(useResidualWeights);
+					removeSeasonality();
+					updateSeasonalAndTrend(useResidualWeights);
+				}
+			}
+			else {
+				for (int iteration = 0; iteration < fInnerIterations; ++iteration) {
+					removeSeasonality();
+					updateSeasonalAndTrend(useResidualWeights);
+					smoothSeasonalSubCycles(useResidualWeights);
+				}
```

Comment:
In the first iteration, removing the seasonality is, in reality, skipped as it operates on a zero vector. Therefore, the initiation from the trend and exog. smoothing part rather than the seasonality smoothing removes more dominant/irregular effects first, which was observed to perform better.